primitives/src/subset.rs

```diff
@@ -0,0 +1,133 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+use sp_std::prelude::Vec;
+
+/// Ordering in a lattice, for example, the lattice of subsets of a set.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum LatticeOrdering {
+    /// Inclusion of the first subset into the second subset.
+    Less,
+    /// Set equality.
+    Equal,
+    /// The subsets are pairwise different.
+    Incomparable,
+    /// Inclusion of the second subset into the first subset.
+    Greater,
+}
+
+/// The lattice order trait.
+pub trait LatticeOrd {
+    /// The lattice comparison.
+    fn lattice_cmp(&self, other: &Self) -> LatticeOrdering;
+}
+
+/// The type of subsets of an open set of elements of type `A` where the whole set is always
+/// considered to be bigger than any finite set of its elements. This is true for infinite
+/// sets. When talking about finite sets, we have to add that they are _open_.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Subset<A> {
+    /// The set of all elements.
+    All,
+    /// A subset of given elements. It is strictly contained in [`Subset::All`].
+    Elems(Vec<A>),
```

Comment:
Instead of having a vector of elements, maybe we can have a BTreeSet of elements and then use the inbuilt `is_subset`, `is_superset` functions for ordering?

Reply:
Indeed! I forgot that `BTreeSet` was added into `sp_std`.